### PR TITLE
Add windows defines for Windows DLL version - libusb-win32 project

### PR DESCRIPTION
**Pull request by rsandoz**
(1) Update (literal array terminator):
```
// C2059: syntax error: '}'
#if defined(WIN32) || defined(_WIN64)
{ 0 }
#else
{ }
#endif
```
Applied to:
```
static const struct hid_device_id razer_devices[] = {
static const struct razer_key_translation chroma_keys[] = {
```
in file(s):
```
razer*_driver.c
```

(2) Update (static for duplicate functions):
```
//LNK2005 already defined in razer*_driver.obj
#if defined(WIN32) || defined(_WIN64)
static
#endif
```
Applied to:
```
struct razer_report razer_send_payload(
int razer_get_report(
void razer_set_device_mode(
```
in file(s):
```
razer*_driver.c
```

(3) Update (union cast):
```
//C2440 'type cast': cannot convert from 'unsigned char' to
'razer_kraken_effect_byte'
#if defined(WIN32) || defined(_WIN64)
unsigned char effect_byte1 = get_current_effect(dev);
union razer_kraken_effect_byte effect_byte;
memcpy(&effect_byte, &effect_byte1, sizeof(unsigned char));
#else
 union razer_kraken_effect_byte effect_byte = (union
razer_kraken_effect_byte)get_current_effect(dev);
#endif
```
Applied to:
```
static ssize_t razer_attr_read_mode_breath(
```
in file(s):
```
razerkraken_driver.c
```

(4) Update (device macro repurpose to DLL API calls):
```
#if defined(WIN32) || defined(_WIN64)
#undef DEVICE_ATTR
#define DEVICE_ATTR(_name, _mode, _show, _store) DEVICE_ATTR1(*, _name,
_mode, _show, _store)
#endif
```
Applied to:
```
static DEVICE_ATTR(
```
in file(s):
```
razer*_driver.c
```

### CODE REVIEW

**Comment by terricain** [comment]
With this does it then build on windows. if said visual studio support crap is present? Or would more mods be needed?

**Comment by z3ntu** [comment]
Yeah how do you build the DLL?

**Comment by terricain** [comment]
Stick a readme in the winproject folder detailing what people need to build this like probs libusb and visualstudio and possibly a short desc of the example code in there initially

**Comment by rsandoz** [comment]
I added a subfolder with all of the VS support stuff.  Basically, it contains the vs project files and all of the linux headers with enough stucts, macros, defines with the preserved linux definitions, but to execute the same functionality in  the windows world (including USB send control message).

The only modifications to your files are what I outlined in my comments which check for windows defines.

**Comment by rsandoz** [comment]
I really have no love for windows.  I am really just hoping that this increases the device support for your drivers in general.

**Comment by rsandoz** [comment]
I will do the readme.  Also, I took this a step back and attempted on an untouched box.  I added in the needed libusb libs in the winproject subfolder.  After building the DLL, I found LoadLibrary to be returning null without displaying any error -- still trying to resolve this.

**Comment by rsandoz** [comment]
I would also like to redo the default directory structure that Visual Studio builds the DLLs in to be cleaner.  the Debug,Release,x86,x64 seems messy and inconsistent.

Also, even after you press the Clean option in Visual Studio it seems to leave a lot of unnecessary items around (*.db, *.filters, *.users, a hidden .vs folder)

**Comment by rsandoz** [comment]
FYI, I also discovered that one of my other devices, my old DeathStalker Ultimate (aka #define USB_DEVICE_ID_RAZER_DEATHSTALKER_ULTIMATE 0x0114) works within your razerkbd_driver.h,c.  It has all the same functions as USB_DEVICE_ID_RAZER_ANANSI

**Comment by terricain** [comment]
> FYI, I also discovered that one of my other devices, my old DeathStalker Ultimate

Nice, make an issue for it otherwise i'll forget but good, I'll add 0x0114 and essentially clone the anansi
Am working on deasthstalker expert as well

**Comment by rsandoz** [comment]
> Stick a readme in the winproject folder detailing what people need to build this like probs libusb and visualstudio and possibly a short desc of the example code in there initially

Done.

**Comment by rsandoz** [comment]
I still plan to test (and develop) more this weekend.  I have some untainted computers I want to test the whole setup procedure on.

**Comment by z3ntu** [comment]
Please don't put binaries into the repo (especially not 5MB files)!

**Comment by rsandoz** [comment]
WIll do.  Will remove any tools binaries as well as the pre-built DLL's and maintain these in a separate 'binaries/release' repo on my end.

**Comment by terricain** [comment]
Am going to be doing some restructuring eventually. What I was thinking was linking the windows bit to another repo (possibly yours / one you have r/w to) as as subrepo, for now it should be fine as it is though

**Comment by terricain** [comment]
Will merge eventually, theres some other stuff i want to get in first though

**Comment by rsandoz** [comment]
I manually created a new repo on my account (razer-drivers-win32) and moved the winproject files into it.

I just followed this submodule instruction:
https://git-scm.com/book/en/v2/Git-Tools-Submodules

And tested as follows:
delete local content
git clone --recursive https://github.com/rsandoz/razer-drivers.git

Also added support for some older Razer devices I had lying around.
USB_DEVICE_ID_RAZER_BLADE_2015
USB_DEVICE_ID_RAZER_DEATHSTALKER_ULTIMATE
USB_DEVICE_ID_RAZER_NAGA_EPIC_WIRELESS
USB_DEVICE_ID_RAZER_NAGA_EPIC_WIRED
USB_DEVICE_ID_RAZER_OROCHI_2013

FYI,
I tried some instructions I found for git 'moving subdir to submodule':
filter-branch (newer?)
https://help.github.com/articles/splitting-a-subfolder-out-into-a-new-repository/
subtree split (older?)
http://stackoverflow.com/questions/359424/detach-move-subdirectory-into-separate-git-repository/17864475#17864475

both of which ended for me in the middle somewhere with an obscure error that I got frustrated with and didn't care to research and resolve

So, I just did what I knew should work.

**Comment by terricain** [comment]
> Also added support for some older Razer devices I had lying around.

By added support for some razer devices, i take it you tested all the options on all of them?
^^ sorry but have to ask

**Comment by rsandoz** [comment]
All of the lighting options using my win32 port.  I'll try and boot  into my linux later on in the week and test.  I really added them more just to test multiple keyboards and multiple mice were being opened proplerly.  Actually I tested when under libusb-win32.  I will run a quick test against my winusb and get back in about an hour (need to go collect my devices).

**Comment by rsandoz** [comment]
I actually found 1 bug (only in my subproject code in hid.h where I was only opening the first device of a kind (1st keyboard, 1st mouse).  I moved SetupDiGetClassDevs into the device loop and that resolved the issue.  I then retested successfully on both keyboards simultaneously and got the expected result.  I'll sync up in a bit.   Is it just me or do you have to use the command line when updating a project with subprojects?

**Comment by rsandoz** [comment]
I actually have a logitech G900 mouse and have capped the "changing of the light" with wireshark.  I wonder if I should put that in?  although it may be entirely too different as it seems like it was using a different USB message.

**Comment by rsandoz** [comment]
Updated.  Change is in.  Tested with new devices.

**Comment by terricain** [comment]
Nice. Am going to wait on this pr as am going to me moving round the repos to the new oprnrazer github organisation so will do the needed messing around then. I also think I can get away with marking a few methods as static and that'll do away with some of the ifdef windoze

**Comment by terricain** [comment]
Pr is looking good, will see if i can cut down on the ifdefs this weekend and then look into merging it

**Comment by rsandoz** [comment]
I took out the legacy devices for now.  Will put them back later on in 
individual pull requests and try to test them more thoroughly per 
request, that way, I can hit the python/Linux upper dependencies as well.

File: driver/razercore_driver.c [hunk 1]
```diff
@@ -577,7 +581,12 @@ static void razer_core_disconnect(struct hid_device *hdev)
  */
 static const struct hid_device_id razer_devices[] = {
     { HID_USB_DEVICE(USB_VENDOR_ID_RAZER,USB_DEVICE_ID_RAZER_CORE) },
+// C2059: syntax error: '}'
+#if defined(WIN32) || defined(_WIN64)
+    { 0 }
```
Comment (terricain) [inline driver/razercore_driver.c:586]: Think this will just work on linux too, it compiles at least.

File: driver/razercore_driver.c [hunk 2]
```diff
@@ -455,6 +455,10 @@ static ssize_t razer_attr_read_device_mode(struct device *dev, struct device_att
  * Write only is 0220
  * Read and write is 0664
  */
+#if defined(WIN32) || defined(_WIN64)
+#undef DEVICE_ATTR
```
Comment (terricain) [inline driver/razercore_driver.c:459]: Would it not make sense to define this somewhere like razercommon.h? then it would only need to be done once right?

**Comment by rsandoz** [comment]
Would that work?

My intent was to undef and redef that macro specific to each different 
driver (firefly, core, kbd, etc)

All I did was re-purpose this macro for the DLL exports. (kind of a hack)

razerfirefly_driver.c
#if defined(WIN32) || defined(_WIN64)
#undef DEVICE_ATTR
#define DEVICE_ATTR(_name, _mode, _show, _store) DEVICE_ATTR1(firefly, 
_name, _mode, _show, _store)
#endif
static DEVICE_ATTR(firmware_version,        0440, 
razer_attr_read_get_firmware_version, NULL);
static DEVICE_ATTR(device_type,             0440, 
razer_attr_read_device_type,          NULL);
...

razercore_driver.c
#if defined(WIN32) || defined(_WIN64)
#undef DEVICE_ATTR
#define DEVICE_ATTR(_name, _mode, _show, _store) DEVICE_ATTR1(core, 
_name, _mode, _show, _store)
#endif
static DEVICE_ATTR(firmware_version,        0440, 
razer_attr_read_get_firmware_version, NULL);
static DEVICE_ATTR(device_type,             0440, 
razer_attr_read_device_type,          NULL);
...

**Comment by terricain** [comment]
Ahh yeah sorry wasnt paying 100% attention, your correct

**Pull request closed without merging**